oslabs-beta/watchmo
# Config commands

With this update, you can view and adjust the config.json files right from the command line!

**What's new**
1. Watchmo configure has many more options available:
- An --endpoint/-e option which allows you to set the endpoint of the given project.
- A --category/-c option which, by default, creates a new category in the config.json file, but also is used in conjunction with other options to specify a category for further manipulation
- A --query/-q option which, by default, creates a new query in the specified category 
- A --frequency/-f option which allows you to change
- And a --remove/-r flag which allows you to remove queries and categories from a given project
- A --view/-v option which prints out the config file to the console

2. Watchmo, the default command, now prints the logo when ran by itself. Perhaps in the future we can have data visualization in the terminal be printed this way. Additionally, watchmo --view prints out the current active projects watchmo has files for.

3. Additionally, the test suite has been improved, including tests for some of the options. --query and --frequency remain untested, currently.

**How to test it**
Run `npm test cli`. We now have a robust suite of unit tests to check the CLI functionality!

**How it works**
Yargs does all the work. In the commands/configure.js file, the configure function acts as a router to route the input flags to the appropriate function.

For each command, run `watchmo configure [project name] (options)`, where the options are as follows: 
- `--endpoint` (alias -e): takes a positional argument that sets the endpoint
- `--category` (alias -c): takes a positional argument that names a new category for [project name] config.json, and writes a template object as the value corresponding to the category name. If used in conjunction with --frequency or --query, this specifies the category for which the given frequency or query is to be adjusted. If used in conjunction with the --remove flag, the entire category is removed from the config.json file.
The template for a new category is given as follows:
```
// watchmoData/projectName/config.json
{ 
  endpoint: 
  categories: {
    // if you ran watchmo configure projectName --category categoryName, this is what would be created:
    categoryName: {  
      queries: [],
      frequency: -1,
    }
  }
}
```
- `--query` (alias -q): takes a positional argument that, when used with --category [category name], adds the argument as a query to the specified category. When the --remove flag is also added, --query takes in an _integer specifying the index of the query_ from the given category to be removed. These integers are easily viewed with `watchmo configure --view`.
- `--frequency` (alias -f): takes a positional argument that, when used with --category [category name], adds the argument as the new frequency for the specified category. 
- `--remove` (alias -r): a flag used in conjunction with either --category by itself, or --category and --query together. e.g. `watchmo configure projectName --category catName --query qIndex --remove` and `watchmo configure projectName --category catName --remove` are valid commands. `watchmo configure projectName --query qIndex` is not.
- `--view` (alis -v): an option that works only on its own, which prints out the associated config.json file. i.e., `watchmo configure projectName --view` is the only way this command should be used.

Note: These commands must run one at a time. When trying to develop functionality to run multiple configure options at once, a bug occurred where the entire config.json file was deleted. Currently the code is written such that any attempt to configure multiple pieces at once (such as changing the endpoint and the frequency) will result in an error message rather than running any function at all.

- `watchmo --view` prints out the list of projects in the src/watchmoData/projectNames.json file.

**Example Workflow**
1. Run `npm test` to make sure everything is up and running.
2. Run `watchmo configure newProject`. You should see a newProject file created in src/watchmoData/.
3. Run `watchmo configure newProject --view`. You should see an unconfigured config file printed to the console. This step should be repeated after each of the following steps to view how the commands are changing the config.json file. Note that an equivalent command is `watchmo configure newProject -v`.
4. Run `watchmo configure --endpoint newEndpoint`. This should change the endpoint to 'newEndpoint'.
5. Run `watchmo configure --category newCategory`. This should add a category called newCategory.
6. Run `watchmo configure --category newCategory --query newQuery`. This should add a query called newQuery to the category newCategory.
7. Run `watchmo configure --category newCategory --frequency 5000`. This should change the frequency in newCategory to 5000.
8. Run `watchmo configure --category newCategory --query 0 --remove`. This should remove the 0th query from the queries in newCategory.
9. Run `watchmo configure --category newCategory --remove`. This should remove the category newCategory.
